File: scripts/ollama_client.py
"""
Wrapper para interactuar con Ollama localmente.
Compatible con OpenAI SDK syntax para fácil integración.
"""
import requests
import json
from typing import Optional, List, Dict, Any


class OllamaClient:
    """Cliente para usar modelos locales con Ollama."""

    def __init__(self, base_url: str = "http://localhost:11434", model: str = "mistral:latest"):
        self.base_url = base_url
        self.model = model
        self.api_url = f"{base_url}/api"

    def chat(self, messages: List[Dict[str, str]], **kwargs) -> Dict[str, Any]:
        """
        Enviar prompt a Ollama compatible con OpenAI SDK.

        Args:
            messages: Lista de mensajes [{"role": "user", "content": "..."}]
            **kwargs: stream, temperature, top_p, etc.

        Returns:
            Respuesta en formato compatible con OpenAI
        """
        endpoint = f"{self.api_url}/chat"

        payload = {
            "model": self.model,
            "messages": messages,
            "stream": kwargs.get("stream", False),
            "temperature": kwargs.get("temperature", 0.7),
            "top_p": kwargs.get("top_p", 0.9),
        }

        try:
            if kwargs.get("stream"):
                return self._stream_chat(endpoint, payload)
            else:
                response = requests.post(endpoint, json=payload, timeout=300)
                response.raise_for_status()
                data = response.json()

                # Convertir a formato OpenAI compatible
                return {
                    "id": "ollama-" + str(hash(str(messages))),
                    "object": "chat.completion",
                    "created": 0,
                    "model": self.model,
                    "choices": [{
                        "index": 0,
                        "message": {
                            "role": "assistant",
                            "content": data.get("message", {}).get("content", "")
                        },
                        "finish_reason": "stop"
                    }],
                    "usage": {
                        "prompt_tokens": 0,
                        "completion_tokens": 0,
                        "total_tokens": 0
                    }
                }
        except Exception as e:
            raise Exception(f"Error en Ollama: {e}")

    def _stream_chat(self, endpoint: str, payload: dict):
        """Streaming responses desde Ollama."""
        response = requests.post(endpoint, json=payload, stream=True, timeout=300)
        for line in response.iter_lines():
            if line:
                yield json.loads(line)

    def list_models(self) -> List[str]:
        """Listar modelos disponibles."""
        try:
            response = requests.get(f"{self.api_url}/tags")
            response.raise_for_status()
            data = response.json()
            return [m["name"] for m in data.get("models", [])]
        except Exception as e:
            raise Exception(f"Error listando modelos: {e}")

    def pull_model(self, model_name: str) -> None:
        """Descargar un modelo."""
        endpoint = f"{self.api_url}/pull"
        payload = {"name": model_name}

        try:
            response = requests.post(endpoint, json=payload, stream=True, timeout=3600)
            for line in response.iter_lines():
                if line:
                    data = json.loads(line)
                    print(f"📦 {data.get('status', 'Descargando...')}")
        except Exception as e:
            raise Exception(f"Error descargando modelo: {e}")


# Alias compatible con OpenAI
def create_client(model: str = "mistral:latest") -> OllamaClient:
    """Factory function para crear cliente Ollama."""
    return OllamaClient(model=model)


if __name__ == "__main__":
    # Ejemplo de uso
    client = OllamaClient(model="mistral:latest")

    # Listar modelos
    print("📦 Modelos disponibles:")
    for model in client.list_models():
        print(f"  • {model}")

    # Chat simple
    print("\n🤖 Chat con Mistral:")
    response = client.chat([
        {"role": "user", "content": "¿Cuál es 2+2?"}
    ])
    print(response["choices"][0]["message"]["content"])

File: scripts/ollama_client.js
/**
 * Cliente Ollama para JavaScript/Node.js
 * Compatible con fetch y Axios
 */

class OllamaClient {
  constructor(baseUrl = "http://localhost:11434", model = "mistral:latest") {
    this.baseUrl = baseUrl;
    this.model = model;
    this.apiUrl = `${baseUrl}/api`;
  }

  /**
   * Enviar chat a Ollama
   * @param {Array<{role: string, content: string}>} messages
   * @param {Object} options - temperatura, top_p, etc
   * @returns {Promise<Object>} Respuesta compatible con OpenAI
   */
  async chat(messages, options = {}) {
    const endpoint = `${this.apiUrl}/chat`;

    const payload = {
      model: this.model,
      messages,
      stream: options.stream || false,
      temperature: options.temperature || 0.7,
      top_p: options.top_p || 0.9,
    };

    try {
      const response = await fetch(endpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        throw new Error(`Ollama error: ${response.statusText}`);
      }

      const data = await response.json();

      // Formato compatible con OpenAI
      return {
        id: `ollama-${Date.now()}`,
        object: "chat.completion",
        created: Math.floor(Date.now() / 1000),
        model: this.model,
        choices: [
          {
            index: 0,
            message: {
              role: "assistant",
              content: data.message?.content || "",
            },
            finish_reason: "stop",
          },
        ],
        usage: {
          prompt_tokens: 0,
          completion_tokens: 0,
          total_tokens: 0,
        },
      };
    } catch (error) {
      throw new Error(`Error en Ollama: ${error.message}`);
    }
  }

  /**
   * Streaming chat
   * @param {Array<{role: string, content: string}>} messages
   * @param {Function} onData - Callback para cada chunk
   */
  async chatStream(messages, onData) {
    const endpoint = `${this.apiUrl}/chat`;

    const payload = {
      model: this.model,
      messages,
      stream: true,
    };

    try {
      const response = await fetch(endpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        throw new Error(`Ollama error: ${response.statusText}`);
      }

      const reader = response.body.getReader();
      const decoder = new TextDecoder();

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = decoder.decode(value);
        const lines = chunk.split("\n").filter((l) => l.trim());

        for (const line of lines) {
          try {
            const data = JSON.parse(line);
            onData(data);
          } catch (e) {
            // Ignorar líneas inválidas
          }
        }
      }
    } catch (error) {
      throw new Error(`Error en Ollama stream: ${error.message}`);
    }
  }

  /**
   * Listar modelos disponibles
   * @returns {Promise<Array<string>>}
   */
  async listModels() {
    try {
      const response = await fetch(`${this.apiUrl}/tags`);
      if (!response.ok) throw new Error("Error listando modelos");

      const data = await response.json();
      return data.models?.map((m) => m.name) || [];
    } catch (error) {
      throw new Error(`Error en listModels: ${error.message}`);
    }
  }

  /**
   * Descargar modelo
   * @param {string} modelName
   */
  async pullModel(modelName) {
    const endpoint = `${this.apiUrl}/pull`;

    try {
      const response = await fetch(endpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ name: modelName }),
      });

      if (!response.ok) throw new Error("Error descargando modelo");

      const reader = response.body.getReader();
      const decoder = new TextDecoder();

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        const chunk = decoder.decode(value);
        const lines = chunk.split("\n").filter((l) => l.trim());

        for (const line of lines) {
          try {
            const data = JSON.parse(line);
            console.log(`📦 ${data.status || "Descargando..."}`);
          } catch (e) {
            // Ignorar
          }
        }
      }
    } catch (error) {
      throw new Error(`Error en pullModel: ${error.message}`);
    }
  }
}

// Exportar
module.exports = { OllamaClient };

// Ejemplo de uso
if (require.main === module) {
  (async () => {
    const client = new OllamaClient("http://localhost:11434", "mistral:latest");

    // Listar modelos
    console.log("📦 Modelos disponibles:");
    const models = await client.listModels();
    models.forEach((m) => console.log(`  • ${m}`));

    // Chat
    console.log("\n🤖 Chat:");
    const response = await client.chat([
      { role: "user", content: "¿Cuál es 2+2?" },
    ]);
    console.log(response.choices[0].message.content);
  })();
}
